src/main/java/com/google/devtools/build/lib/remote/grpc/SharedConnectionFactory.java

```diff
@@ -155,5 +183,11 @@ public void close() throws IOException {
     public Connection getUnderlyingConnection() {
       return connection;
     }
+
+    private static boolean isFatalError(@Nullable Throwable t) {
+      // A low-level netty error indicates that the connection is fundamentally broken
+      // and should not be reused for retries.
+      return t instanceof Errors.NativeIoException;
```

Comment:
We could extend this to cover certain status codes, but I wanted to start with something limited enough to not introduce performance regressions.